### ISSUE TITLE
SPACK_SHELL set incorrectly within cobalt scheduler job script environment

### ISSUE THREAD

**Issue by carns** [spack/spack#8443]
I have been unable to get any spack load commands to work within a cobalt job script even though they work fine on a login node or within an interactive job session.  This is on the "cooley" system at the ALCF (https://www.alcf.anl.gov/user-guides/cooley).

After some debugging, I found that this is because the SPACK_SHELL variable is being set incorrectly.  This is in my env after running setup-env.sh:

```SPACK_SHELL=margo-regressio```

("margo-regressio" is a truncation of the actual script name).  The SPACK_SHELL variable is generated from this logic:

https://github.com/spack/spack/blob/develop/share/spack/setup-env.sh#L201

If I check the output of ```ps -p $$``` within a job script, I get the following:

```
ps -p 106371
   PID TTY          TIME CMD
106371 ?        00:00:00 margo-regressio
```

Note that the CMD is "margo-regressio" intead of "bash" like the setup-env.sh logic expects.  If I manually edit setup-env.sh to force SPACK_SHELL to be set to "bash", then everything works as expected for me.

Is there a more robust way to detect the shell type in setup-env.sh, or maybe even just a way to override it from the command line?

**Comment by carns** [comment]
I think this happens on any Linux machine of you execute a shell script directly:

```
[carns@cooleylogin2 mochi-regression-job-3451395]$ ./foo.sh 
    PID TTY          TIME CMD
3857456 pts/61   00:00:00 foo.sh
[carns@cooleylogin2 mochi-regression-job-3451395]$ cat foo.sh
#!/bin/bash
ps -p $$
```

Is there something else that I should use inside of a script besides setup-env.sh?  Or is this supposed to work and we need different logic to detect the shell type?

**Comment by carns** [comment]
This seems to work for me:

```[carns@cooleylogin2 mochi-regression-job-3451395]$ ./foo.sh 
bash
[carns@cooleylogin2 mochi-regression-job-3451395]$ cat foo.sh 
#!/bin/bash
basename $SHELL
```

I could submit a PR to make this change but I'm curious if I'm even barking up the right tree here :)

**Comment by adamjstewart** [comment]
> This is in my env after running setup-env.sh

Are you _running_ the script or _sourcing_ it? You should source it in your job submission script.

> or maybe even just a way to override it from the command line?

You can test out #7475. It lets you override it by setting the environment variable yourself.

`$SHELL` works most of the time, but doesn't always work. Specifically, `$SHELL` is the login shell, not the current shell. If you login using `bash` and switch to `zsh`, it will still say `bash`. The `ps` logic is supposed to be more robust, but as you've discovered, nothing is perfect.

P.S. See https://stackoverflow.com/a/3327022/5828163 for a more in depth analysis of the various options.

**Comment by carns** [comment]
Thanks for the pointer @adamjstewart , I'll try out #7475 and report back!

I was indeed sourcing the script rather than running it, just a little careless in my description :)

**Comment by carns** [comment]
#7475 indeed worked just fine if I explicitly set SPACK_SHELL before sourcing setup-env.sh.  Thanks!

I'm also closing this issue even though the PR hasn't landed since it seems to be a duplicate of #7441.  I'll just watch that issue for updates.
